server-spi-private/src/main/java/org/keycloak/models/dblock/DBLockProvider.java

```diff
@@ -27,12 +27,30 @@
  */
 public interface DBLockProvider extends Provider {
 
+    /**
+     * Lock namespace to have different lock types or contexts.
+     */
+    public enum Namespace {
+
+        GENERAL(1),
+        OFFLINE_SESSIONS(1000);
+
+        private final int id;
+
+        private Namespace(int id) {
+            this.id = id;
+        }
+
+        public int getId() {
+            return id;
+        }
+    };
 
     /**
      * Try to retrieve DB lock or wait if retrieve was unsuccessful. Throw exception if lock can't be retrieved within specified timeout (900 seconds by default)
+     * @param lock The lock namespace to use
      */
-    void waitForLock();
-
+    void waitForLock(Namespace lock);
```

Comment:
It seems a bit unclear that waitForLock has namespace argument, but for the other methods like "releaseLock" or "hasLock" it is not very clear how they work now.

It seems that with the liquibase implementation, you can't have "nested" lock (EG. obtain lock for namespace GENERAL and then have same node to obtain the lock also for OFFLINE_SESSIONS namespace when GENERAL is still held). Is it possible to make it clear in the javadoc? Also is it possible to rename the "hasLock" to "getCurrentLock", which will return null or namespace of the currently retrieved lock? For "releaseLock", I guess it will be good to not change signature, but just add the information that it is used to release the lock, which is currently held. WDYT?

Reply:
@mposolda I didn't think about nested locks... But I think that the "hasLock" method has the meaning if this node has the lock (previously the only one that was possible). So it should be "hasLock(Namespace)" for example is used [here](https://github.com/keycloak/keycloak/blob/master/model/jpa/src/main/java/org/keycloak/connections/jpa/DefaultJpaConnectionProviderFactory.java#L330) without specifying any "waitForLock" before. Maybe it's better if we add the Namespace to the three methods (waitForLock, releaseLock and hasLock). WDYT?

Reply:
No, the "releaseLock" should be done for the previous obtained lock. So the summary is: "waitForLock" and "hasLock" receive the namespace (the first method performs the select for update to the ID, the second just obtains if there is a lock from a concurrent set in the provider); the "releaseLock" rollbacks the previous select for update. When you get a "DBLockProvider" you can only waitForLock/releaseLock one namespace, if you need nested you need two providers (two connections).

I'm going to prepare the changes, let's see if I can commit something the next week.